### PR TITLE
Adding ClearGridExceptRegion service (#220)

### PR DESCRIPTION
**Pull request by ghost**
Hello!

As we discussed on #220, I've created a service for clearing the grid except for a safety square region around the robot. The region is defined with the parameter reset_distance that is the side of the square, as its also defined in the ClearExceptRegion on navigation2 side.

I won't say I'm 100% happy with this implementation, as I had to do something ugly to get access to robot position when doing the clearing because we don't have access to it at all times (hence when we call the service (I think)). But this is working for us!

So we've decided to open this PR anyways, see if it's of your interest to merge it, or if there are better ideas for the implementation before the merge.

### CODE REVIEW

File: src/spatio_temporal_voxel_grid.cpp [hunk 1]
```diff
@@ -451,4 +451,51 @@ bool SpatioTemporalVoxelGrid::SaveGrid(
...
+      if (this->ClearGridPoint(pt_index)) {
+        cleared_cells.insert(occupany_cell(x, y));
+      }
+      else {
+        std::cout << "Failed to clear point." << std::endl;
+      }
+    }
+    else {
+      // if here, we can add to costmap and PC2
+      PopulateCostmapAndPointcloud(pt_index);
```
Comment (SteveMacenski) [inline src/spatio_temporal_voxel_grid.cpp]: Would this not already be happening via other mechanics?

Comment (ghost) [reply src/spatio_temporal_voxel_grid.cpp]: That's true! it's done with the clearings on TemporalClearAndGenerateCostmap. Changed it so we only do it there, and also added restriction so it's only possible to clear when not on mapping mode, as it is for the frustum/temporal clearings.

File: src/spatio_temporal_voxel_layer.cpp [hunk 2]
```diff
@@ -768,6 +776,39 @@ void SpatioTemporalVoxelLayer::SaveGridCallback(
...
+}
+
+/*****************************************************************************/
+void SpatioTemporalVoxelLayer::ClearGridExceptRegion(
+  double robot_x, double robot_y,
+  std::unordered_set<volume_grid::occupany_cell> & cleared_cells)
+/*****************************************************************************/
+{
+  if (!_clear_grid_except_region)
+    return;
```
Comment (SteveMacenski) [inline src/spatio_temporal_voxel_layer.cpp:806]: Add brackets in the if condition, its a 1 liners so I know you "can" but if we expanded that later I've had problems where developers forget to later add it

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Last small change and good to go, I assume you tested this?

File: src/spatio_temporal_voxel_layer.cpp (hunk 2 above)
Comment (ghost) [reply src/spatio_temporal_voxel_layer.cpp:806]: Right, that's understandable. Added the brackets there. 

Yes have been testing and re-testing this after the changes!

**Review by SteveMacenski — APPROVED** [review]
This looks fine to me. If this were to be run often, I would rearchitect this so that we implement the local clearing in the actual ClearFrustum API so that we only iterate through the grid exactly 1 time, but since this is irregular and infrequent I think this is OK. If you do notice performance issues, that would be the low-hanging fruit solution. It would involve changing the API a little bit, but that way you can leverage the same iteration / world coordinate conversion that's already happening. 

its also probably possible in the openvdb API to define a range of values you're interested in and iterate only through those, but again, infrequently this is probably OK, even if suboptimal

**Pull request merged**